Gems/EMotionFX/Code/EMotionFX/Source/MotionLayerSystem.cpp

```diff
@@ -157,10 +157,7 @@ namespace EMotionFX
                 if (source->GetBlendOutBeforeEnded())
                 {
                     // if the motion has to stop
-                    if ((!source->GetIsPlayingForever() &&
-                        (source->GetNumCurrentLoops() >= source->GetMaxLoops()) &&
-                        (source->GetTimeDifToLoopPoint() <= source->GetFadeTime()) &&
-                        !source->GetFreezeAtLastFrame()) ||
+                    if ((!source->GetIsPlayingForever() && source->GetTimeDifToLoopPoint() <= source->GetFadeTime() && !source->GetFreezeAtLastFrame()) ||
```

Comment:
NIT: This is a bit difficult to read, I personally don't like super long if statement.
I'd rather have something like this:
bool motionHasToStop = false;
if ((!source->GetIsPlayingForever() && source->GetTimeDifToLoopPoint() <= source->GetFadeTime() && !source->GetFreezeAtLastFrame())
{ motionHasToStop = true; }
else if(source->GetHasEnded() && !source->GetFreezeAtLastFrame()))
{ motionHasToStop = true; }

then if(motionHasToStop)
{
...
}

Reply:
Simplified it and fixed the tests (they were checking something wrong). Also uploaded a video demonstrating all the modes and settings working correctly.